marshal/marshal.go

```diff
@@ -0,0 +1,212 @@
+		name: parts[0],
+	}
+
+	for _, v := range parts {
+		switch v {
+		case "omitempty":
+			jsonTag.omitempty = true
+		case "string":
+			jsonTag.asString = true
+		}
+	}
+	return jsonTag, nil
+}
+
+// Determines whether a field should be included or not
+func includeField(tag jsonTag, value reflect.Value, forceSend bool) bool {
+	if tag.ignore {
+		return false
+	}
+	if !tag.omitempty {
+		return true
+	}
+	return (isBasicType(value.Type()) && forceSend) || !isEmptyValue(value)
+}
+
+// isEmptyValue returns whether v is the empty value for its type.
+// This implementation is based on on the encoding/json package for consistency on the results
+// https://github.com/golang/go/blob/a278550c40ef3f01a5fcbef43414dc49009201f8/src/encoding/json/encode.go#L306
+func isEmptyValue(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
+		return v.Len() == 0
+	case reflect.Bool:
+		return !v.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return v.Int() == 0
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return v.Uint() == 0
+	case reflect.Float32, reflect.Float64:
+		return v.Float() == 0
+	case reflect.Interface, reflect.Pointer:
+		return v.IsNil()
+	}
+	return false
+}
+
+// Merges two maps. Overrides duplicated elements, which is fine because json.Unmarshal also
+// does it, and a JSON should not have duplicated entries.
+func mergeMaps(m1 map[string]any, m2 map[string]any) map[string]any {
+	merged := make(map[string]any)
+	maps.Copy(merged, m2)
+	maps.Copy(merged, m1)
+	return merged
+}
+
+func getForceSendFields(v any, structName string) (map[string]bool, error) {
+	// reflect.GetFieldByName panics if the field is inside a null anonymous field
+	field := getFieldByName(v, force_send_field_name)
+	if !field.IsValid() {
+		return nil, nil
+	}
+	forceSendFields, ok := field.Interface().([]string)
+	if !ok {
+		return nil, fmt.Errorf("invalid type for %s field", force_send_field_name)
+	}
+	existingFields := getFieldNames(v)
+	includeFields := make(map[string]bool)
+	for _, field := range forceSendFields {
+		if _, ok := existingFields[field]; ok {
+			includeFields[field] = true
+		} else {
+			return nil, fmt.Errorf("field %s cannot be found in struct %s", field, structName)
+		}
+	}
+
+	return includeFields, nil
+
+}
+
+func getFieldByName(v any, fieldName string) reflect.Value {
+	value := reflect.ValueOf(v)
+	value = reflect.Indirect(value)
+	objectType := value.Type()
+
+	for i := 0; i < value.NumField(); i++ {
+		name := objectType.Field(i).Name
+		if name == fieldName {
+			return value.Field(i)
+		}
+	}
+	return reflect.Value{}
+}
+
+func getFieldNames(v any) map[string]bool {
+	result := map[string]bool{}
+	value := reflect.ValueOf(v)
+	value = reflect.Indirect(value)
+	objectType := value.Type()
+
+	for i := 0; i < value.NumField(); i++ {
+		name := objectType.Field(i).Name
```

Comment:
Should this skip anonymous fields?

Reply:
We can skip anything which is not a basic type.
Anonymous fields do have a name internally (which corresponds to the struct type name), so they are not really special here.